src/adapters/sqlite/makeDispatcher/node/DatabaseDriver.js

```diff
@@ -0,0 +1,292 @@
+// @flow
+
+import Database from './Database'
+
+function fixArgs(args: any): any {
+  return Object.keys(args).reduce((acc, argName) => {
+    if (typeof acc[argName] === 'boolean') {
+      acc[argName] = acc[argName] ? 1 : 0
+    }
+    return acc
+  }, args)
+}
+
+type Migrations = { from: number, to: number, sql: string }
+
+class MigrationNeededError extends Error {
+  databaseVersion: number
+
+  type: string
+
+  constructor(databaseVersion): void {
+    super('MigrationNeededError')
+    this.databaseVersion = databaseVersion
+    this.type = 'MigrationNeededError'
+    this.message = 'MigrationNeededError'
+  }
+}
+
+class SchemaNeededError extends Error {
+  type: string
+
+  constructor(): void {
+    super('SchemaNeededError')
+    this.type = 'SchemaNeededError'
+    this.message = 'SchemaNeededError'
+  }
+}
+
+export function getPath(dbName: string): string {
+  if (dbName === ':memory:' || dbName === 'file::memory:') {
+    return dbName
+  }
+
+  let path =
+    dbName.startsWith('/') || dbName.startsWith('file:') ? dbName : `${process.cwd()}/${dbName}`
+  if (path.indexOf('.db') === -1) {
+    if (path.indexOf('?') >= 0) {
+      const index = path.indexOf('?')
+      path = `${path.substring(0, index)}.db${path.substring(index)}`
+    } else {
+      path = `${path}.db`
+    }
+  }
+
+  return path
+}
+
+class DatabaseDriver {
+  static sharedMemoryConnections = {}
+
+  database: Database
+
+  cachedRecords: any = {}
+
+  initialize = (dbName: string, schemaVersion: number) => {
+    this.init(dbName)
+    this.isCompatible(schemaVersion)
+  }
+
+  setUpWithSchema = (dbName: string, schema: string, schemaVersion: number) => {
+    this.init(dbName)
+    this.unsafeResetDatabase({ version: schemaVersion, sql: schema })
+    this.isCompatible(schemaVersion)
+  }
+
+  setUpWithMigrations = (dbName: string, migrations: Migrations) => {
+    this.init(dbName)
+    this.migrate(migrations)
+    this.isCompatible(migrations.to)
+  }
+
+  init = (dbName: string) => {
+    this.database = new Database(getPath(dbName))
+
+    const isSharedMemory = dbName.indexOf('mode=memory') > 0 && dbName.indexOf('cache=shared') > 0
+    if (isSharedMemory) {
+      if (!DatabaseDriver.sharedMemoryConnections[dbName]) {
+        DatabaseDriver.sharedMemoryConnections[dbName] = this.database
+      }
+      this.database = DatabaseDriver.sharedMemoryConnections[dbName]
```

Comment:
does `better-sqlite3` not support shared cache memory mode? I don't believe that this replicates the semantics exactly. But since it passes tests, it's probably not a big deal (since I can't think of many legitimate use cases for using this other than FOR tests), but just saying.

Reply:
So, as far as I spoke to them, no... when in-memory mode, each DB connection seems to be unique. ( https://github.com/JoshuaWise/better-sqlite3/issues/417 )

Reply:
weird, but okay